WordPress/Classes/Utility/Delay.swift

```diff
@@ -0,0 +1,96 @@
+import Foundation
+
+typealias Action = () -> Void
+
+/// Something that can be canceled
+///
+protocol Cancelable {
+    func cancel()
+}
+
+/// A delayed action implemented using GCD
+///
+/// This class will perform the given action after the given delay, unless it is
+/// canceled first.
+///
+/// - Example: Create an action and cancel it.
+///
+///     // Print "action" after 5 seconds
+///     let action = DispatchDelayedAction(delay: 5) { print("action") }
+///     // Actually, cancel
+///     action.cancel()
+///
+struct DispatchDelayedAction: Cancelable {
+    // The whole dance with the cancellation type might seem pointless. After
+    // all, why not use a Bool instead?
+    //
+    // The goal is to have a simple to use API, but one that allows
+    // cancellation.  So you could do something as simple as:
+    //
+    //     DispatchDelayedAction(delay: 5) { doSomething() }
+    //
+    // Any of the Timer-based approaches would require keeping a reference to
+    // self, and even then, the cleaner block-based approach is iOS 10 only.
+    //
+    // For a GCD-based approach, we can just store a canceled flag and prevent
+    // the action from happening. But this would require doing some non-obvious
+    // weak self dance as well. We'd want to capture self weakly, and only
+    // perform the action if self?.canceled is false or self is nil.
+    //
+    // The alternative is to capture a class that contains the cancellation
+    // status instead of self. If we want to keep a reference to self and cancel
+    // that's fine, if we don't the asyncAfter block will capture the
+    // cancellation class instance and release it after it completes.
+
+    /// Performs an action after the given delay, unless it's canceled
+    ///
+    init(delay: DispatchTimeInterval, action: @escaping () -> Void) {
+        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [cancellation] in
+            guard !(cancellation.canceled) else {
+                return
+            }
+            action()
+        }
+    }
+
+    /// Cancels the action if it hasn't been invoked yet
+    ///
+    func cancel() {
+        cancellation.canceled = true
+    }
+
+    private let cancellation = Cancellation(false)
```

Comment:
See [this comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/6357/files#diff-8038861aff8e15fcf63771661f38aa77R24):

The whole dance with the cancellation type might seem pointless. After all, why not use a Bool instead?

The goal is to have a simple to use API, but one that allows cancellation.  So you could do something as simple as:

```swift
DispatchDelayedAction(delay: 5) { doSomething() }
```

Any of the Timer-based approaches would require keeping a reference to self, and even then, the cleaner block-based approach is iOS 10 only.

For a GCD-based approach, we can just store a canceled flag and prevent the action from happening. But this would require doing some non-obvious weak self dance as well. We'd want to capture self weakly, and only perform the action if self?.canceled is false or self is nil.  

The alternative is to capture a class that contains the cancellation status instead of self. If we want to keep a reference to self and cancel that's fine, if we don't the asyncAfter block will capture the cancellation class instance and release it after it completes.